### PR TITLE
test/issues: add failing test for flatten batches out-of-order

### PR DESCRIPTION
**Pull request by adamdecaf**
Issue: https://github.com/moov-io/ach/issues/863

### CODE REVIEW

**Comment by adamdecaf** [comment]
It looks like https://github.com/moov-io/ach/commit/8a3eeb88efcb8cdf25fe7d3f09360c3ee4a544a6 was a partial fix for this issue, but we're still merging/flattening files with out-of-order trace numbers. I think we'll have to re-order those inside `FlattenBatches()`.

**Comment by adamdecaf** [comment]
Actually, should we re-order the entries? That way we aren't modifying the file contents at all. Most people will use the `TraceNumber` as a unique value so our library changing that will break their use-cases.

**Comment by adamdecaf** [comment]
```go
func sortEntries(entries []*EntryDetail) []*EntryDetail {
	sort.Slice(entries[:], func(i, j int) bool {
		return entries[i].TraceNumber < entries[j].TraceNumber
	})
	return entries
}
```

This lets us easily update the entry ordering but with `Batcher` it's not a straightforward usage. I think we can do a quick `b, ok := f.Batches[i].(*Batch); ok` to sort them with `b.Entries = sortEntries(..)`. I'll try that out. Each BatchXXX inherits `Batch` so it _should_ work.

**Comment by adamdecaf** [comment]
@vxio can you give me your thoughts on this?

**Comment by vxio** [comment]
> Actually, should we re-order the entries? That way we aren't modifying the file contents at all. Most people will use the `TraceNumber` as a unique value so our library changing that will break their use-cases.

I think it's fine if we re-order the entries. Similar to recent fixes we made to `MergeFiles(..)`, we have to do some reordering/overriding to ensure items aren't out-of-order or collide.

I like your idea of sorting the entries by traceNumber.

File: entryDetail.go [hunk 1]
```diff
@@ -604,3 +605,10 @@ func (ed *EntryDetail) addendaCount() (n int) {
 	}
 	return n
 }
+
+func sortEntries(entries []*EntryDetail) []*EntryDetail {
```
Comment (vxio) [inline entryDetail.go]:
```suggestion
func sortEntriesByTraceNumber(entries []*EntryDetail) []*EntryDetail {
```

File: file.go [hunk 2]
```diff
@@ -1084,6 +1084,20 @@ func (f *File) FlattenBatches() (*File, error) {
 			}
 		}
 	}
+	for i := range out.Batches {
+		bh := out.Batches[i].GetHeader()
+		if bh.StandardEntryClassCode != "ADV" {
+			batch, _ := NewBatch(bh)
+			entries := sortEntries(out.Batches[i].GetEntries())
```
Comment (vxio) [inline file.go]: We should update the GoDoc to reflect this too

Comment (adamdecaf) [reply file.go]: Fixed, thanks!

**Review by vxio — APPROVED** [review]
Looks good to me!

Should I create an issue to add ADV and IAT support in the near future?

**Comment by adamdecaf** [comment]
Sure. I figure IAT would be requested first. I'm not sure how popular ADV would be for this function.

**Pull request merged**